wallet/core/src/account/pskb.rs

```diff
@@ -0,0 +1,346 @@
+            .map(|utxo_entry| extract_script_pub_key_address(&utxo_entry.script_public_key.clone(), network_id.into()).unwrap()) // Clone the ScriptPublicKey
+            .collect();
+
+        signer.ingest(addresses.as_ref())?;
+
+        // Define the signing function
+        let mut sign = |signer_pskt: PSKT<Signer>| {
+            signer_pskt
+                .pass_signature_sync(|tx, sighash| -> Result<Vec<SignInputOk>, String> {
+                    tx.tx
+                        .inputs
+                        .iter()
+                        .enumerate()
+                        .map(|(idx, _input)| {
+                            let hash = calc_schnorr_signature_hash(&tx.as_verifiable(), idx, sighash[idx], &mut reused_values);
+                            let msg = secp256k1::Message::from_digest_slice(hash.as_bytes().as_slice()).unwrap();
+                            let address = addresses.get(idx).expect("Input indexed address");
+                            let public_key = signer.public_key(address).expect("Public key for input indexed address");
+                            Ok(SignInputOk {
+                                signature: Signature::Schnorr(signer.sign_schnorr(address, msg).unwrap()),
+                                pub_key: public_key,
+                                key_source: None,
+                            })
+                        })
+                        .collect()
+                })
+                .unwrap()
+        };
+
+        signed_bundle.add_pskt(sign(pskt.clone()));
+    }
+    Ok(signed_bundle)
+}
+
+pub fn bundle_to_finalizer_stream(bundle: &Bundle) -> Pin<Box<dyn Stream<Item = Result<PSKT<Finalizer>, Error>> + Send>> {
+    let stream = stream::iter(bundle.inner_list.clone()).map(move |pskt_inner| {
+        let pskt: PSKT<Creator> = PSKT::from(pskt_inner);
+        let pskt_finalizer = pskt.constructor().updater().signer().finalizer();
+
+        let result = pskt_finalizer.finalize_sync(|inner: &Inner| -> Result<Vec<Vec<u8>>, String> {
+            Ok(inner
+                .inputs
+                .iter()
+                .map(|input| -> Vec<u8> {
+                    let signatures: Vec<_> = input
+                        .partial_sigs
+                        .clone()
+                        .into_iter()
+                        .flat_map(|(_, signature)| {
+                            iter::once(OpData65).chain(signature.into_bytes()).chain([input.sighash_type.to_u8()])
+                        })
+                        .collect();
+
+                    signatures
+                        .into_iter()
+                        .chain(
+                            input
+                                .redeem_script
+                                .as_ref()
+                                .map(|redeem_script| ScriptBuilder::new().add_data(redeem_script.as_slice()).unwrap().drain().to_vec())
+                                .unwrap_or_default(),
```

Comment:
this is correct in case of multisig: <signatures>+redeem_script.
in general it depends on script.
for example in case of krc20 it will be one signature + redeem_script.
in case of script True it will be empty signatures + true.

this function must have name/generic explaining when it's useful

Reply:
Renamed to finalize_pskt_one_or_more_sig_and_redeem_script, also works for krc20.

New function added: finalize_pskt_no_sig_and_redeem_script